### PR TITLE
gall: fix bug in +ap-peek

### PR DESCRIPTION
**Pull request by jtobin**
fc7901d2 refactored much of +ap-peek, but introduced a bug in the process.  The relevant diff from that commit is as follows:

```
         =/  slammed
-        =/  =path  [ren tyl]
-        =/  =vase  !>((slag p.u.cug path))
-        (ap-slam q.u.cug p.arm vase)
+        =/  index  p.u.maybe-arm
+        =/  term  q.u.maybe-arm
+        =/  =vase
+          =/  =path  [term tyl]
+          =/  raw  (slag index path)
+          !>  raw
+        (ap-slam term p.arm vase)
```
Note that [ren tyl] was replaced with [term tyl], where 'term' and 'ren' are not equal.  This commit merely rights that wrong.

(/cc @philipcmonk as this is probably still relevant on your static gall branch)

### CODE REVIEW

**Comment by loganallenc** [comment]
Hi Jared, I’m curious as to how the bug was being surfaced. Also, have you tested that the chat app still works? It utilizes +peek extensively.

**Comment by jtobin** [comment]
From @joemfb in an email re: the DNS collector app:

> the read interface (scry/+peek) to :dns-collector is not working. It
> worked last era, so maybe some aspect of app-scry changed in the
> recent %gall refactor?

I took a peek (ha ha, get it) et voilà, there was indeed a shadowing
issue, so *presumably* this caused the regression in :dns-collector.

I hadn't checked the effect of this fix on anything downstream, but I
did boot up Landscape on a fakezod after seeing your comment and clicked
around / did some stuff in chat.  Nothing seemed awry.

It's difficult for me to tell what the effect of the shadowing bug is,
exactly.  If you take a look at the code, it seems like it may produce
some unwanted effect whenever the index is zero -- otherwise the
shadowed term will just be lopped off by `slag`.

**Comment by joemfb** [comment]
Worth noting that this has also broken `:acme`.

**Comment by belisarius222** [comment]
I think new Gall might overuse this kind of shadowing, because it makes this kind of bug a lot more likely.  I was actually worrying about a very similar shadowing in some other Gall arm last night; I think that one is ok, but it is a bit of a footgun.

I wonder if the criterion could be: if you're shadowing a variable with a variable of a different type, it's ok, but if they're the same type, you're opening yourself up to errors.  Of course, that has the problem that if you change the types later and they happen to match up, you could accidentally get back into the bad situation.  There might be a better way of thinking about this.

Maybe we should just try not to shadow variables in general.

**Comment by joemfb** [comment]
I agree that this shadowing feature is being overused. It seems nice in small arms or gates, where the meaning of some value is unambiguously conveyed by the name of its type. I also think a/b/c or p/q/r are fine in similar situations. The shadowing style is worse than ultra-lapidary when it creates a false sense of clarity; and that's easily enough done when you have a couple `+term`'s, for instance.

**Comment by ohAitch** [comment]
Perhaps it would be good to limit use to |= but not =/ then?

**Comment by philipcmonk** [comment]
Well, the usual anti-shadowing measure is =foo=bar, which doesn't work in =/, where it would be most useful.

I think the rule violated here is that "term" is too general for this usage.  Some general principles:

- `=foo` is good if the scope is small, especially less than five lines.
- `=foo` is good if it's specific, like `=care`.  There's likely only one +care you're talking about.  If there's more than one, both should have a more specific name.
- `=foo` is bad if it's too general of a type.  `=term` is rarely a good idea unless you actually only care about it being a term (eg in a few stdlib operations).  `=path` is occasionally useful, but never if you actually mean something more specific like `=wire`.
- If you find yourself using `=general-type`, consider creating another type as a documentary aid.  If a noun, consider adding an aura.  For example: `=term` is bad, `app-name=term` is good, `++  app-name  @tasappname` followed by `=app-name` is best.  `=@app-name` is for those who like to live dangerously.

**Comment by ohAitch** [comment]
Ah did that ever make it in? I thought foo-bar names popping up without
ever being textually declared as such had been deemed too hard to follow

**Comment by joemfb** [comment]
Previously discussed in #1352
